WordPress/gutenberg
# Update to Typescript 5.1

## What?
Update to Typescript 5.1. So far, seems like we need only minimal changes!

The main thing we needed to change was the `importsNotUsedAsValues` rule. This [has been deprecated](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#verbatimmodulesyntax), and we were only using it to enforce `type` imports (see [this comment](https://github.com/WordPress/gutenberg/pull/27143#discussion_r577964719) for why.) [It was recommended](https://github.com/microsoft/TypeScript/pull/52203#issuecomment-1476559412) to use eslint instead if you were only using it to enforce `import type` for type imports. The new option (`verbatimModuleSyntax`) does several other things related to modules which probably won't work for us yet.

To replace this compiler setting, I enabled the `typescript-eslint/consistent-type-imports` rule. This enforces separate type imports and can be autofixed -- I also autofixed all existing reports of this rule. If folks prefer, we could enable the inline type syntax (`import { type Foo } from ...`), which would allow for importing types and modules in the same line.

Additionally, we needed to update `typescript-eslint` from 5.3 to 5.62 for compatibility with TS 5.1. Unfortunately, I [ran into an issue](https://github.com/npm/cli/issues/5840), so we also have to add the two dependencies to our root devDependencies. (For some reason, npm no longer hoists this package after the update.)

## Testing Instructions
- CI should pass
- run `npm run build:package-types` locally
- Verify TS type hints in IDE still works
- Verify that you must use `import type` in ts files for type imports
